modules/key-distributor/cfn/encrypter/function.go

```diff
@@ -0,0 +1,163 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package encrypter
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/armor"
+	"golang.org/x/crypto/openpgp/packet"
+	iam "google.golang.org/api/iam/v1"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"regexp"
+	"strings"
+)
+
+type encryptRequest struct {
+	Email string `json:"email"`
+}
+
+type encryptResponse struct {
+	EncryptedKey string `json:"encryptedKey"`
+}
+
+// Initialize this so the background context persists between invocations.
+var iamService *iam.Service
+
+func init() {
+	ctx := context.Background()
+	var err error
+	if iamService, err = iam.NewService(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func GenerateAndEncrypt(w http.ResponseWriter, r *http.Request) {
+	// Read request bytes
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		e := fmt.Sprintf("Error reading request: %v", err)
+		http.Error(w, e, http.StatusBadRequest)
+		return
+	}
+
+	// Parse the request body to get the Service Account email
+	input := encryptRequest{}
+	if err := json.Unmarshal(data, &input); err != nil {
+		e := fmt.Sprintf("Error parsing request: %v", err)
+		http.Error(w, e, http.StatusBadRequest)
+		return
+	}
+
+	// Read in public key
+	recipient, err := readEntity(os.Getenv("PUBLIC_KEY"))
+	if err != nil {
+		e := fmt.Sprintf("Could not parse public key: %v", err)
+		http.Error(w, e, http.StatusInternalServerError)
+		return
+	}
+
+	// Create service account key
+	b64key, err := createServiceAccountKey(input.Email)
+	if err != nil {
+		e := fmt.Sprintf("Could not create Service Account key: %v", err)
+		http.Error(w, e, http.StatusInternalServerError)
+		return
+	}
+	key, err := base64.URLEncoding.DecodeString(b64key)
+	if err != nil {
+		e := fmt.Sprintf("Could not decode service account key: %v", err)
+		http.Error(w, e, http.StatusInternalServerError)
+		return
+	}
+
+	// Do the encryption
+	var encryptedData bytes.Buffer
+	if err = encrypt([]*openpgp.Entity{recipient}, nil, string(key), &encryptedData); err != nil {
+		e := fmt.Sprintf("Could not encrypt: %v", err)
+		http.Error(w, e, http.StatusInternalServerError)
+		return
+	}
+
+	// Create the response object
+	resp, err := json.Marshal(encryptResponse{
+		EncryptedKey: base64.URLEncoding.EncodeToString(encryptedData.Bytes()),
+	})
+	if err != nil {
+		e := fmt.Sprintf("Could not marshal response: %v", err)
+		http.Error(w, e, http.StatusInternalServerError)
+		return
+	}
+
+	// Write the response
+	io.WriteString(w, string(resp))
+}
+
+func encrypt(recip []*openpgp.Entity, signer *openpgp.Entity, plaintext string, w io.Writer) error {
```

Comment:
Should this accept recipient directly? I'm not sure why it needs a full list of recipients.

Reply:
`openpgp.Encrypt` accepts an array of recipients. We'd have to convert it to a single element array in the `encrypt` method or the `GenerateAndEncrypt` method. I don't really know that one is better than another but if you have strong opinions I can change it :)